src/fromager/sources.py

```diff
@@ -119,6 +140,136 @@ def resolve_source(
+
+    if need_to_clone:
+        with tempfile.TemporaryDirectory() as tmpdir:
+            clone_dir = pathlib.Path(tmpdir) / "src"
+            gitutils.git_clone(
+                ctx=ctx,
+                req=req,
+                output_dir=clone_dir,
+                repo_url=url_to_clone,
+                ref=git_ref,
+            )
+            if not version:
+                # If we still do not have a version, get it from the package
+                # metadata.
+                version = _get_version_from_package_metadata(ctx, req, clone_dir)
+                logger.info("%s: found version %s", req.name, version)
+                working_src_dir = (
+                    ctx.work_dir / f"{req.name}-{version}" / f"{req.name}-{version}"
+                )
+                if working_src_dir.exists():
+                    # We have to check if the destination directory exists
+                    # because if we were not given a version we did not clean it
+                    # up earlier. We do not use ctx.cleanup to control this
+                    # action because we cannot trust that the destination
+                    # directory is reusable because we have had to compute the
+                    # version and we cannot be sure that the version is dynamic
+                    # Two different commits in the repo could have the same
+                    # version if that version is set with static data in the
+                    # repo instead of via a tag or dynamically computed by
+                    # something like setuptools-scm.
+                    logger.debug("%s: cleaning up %s", req.name, working_src_dir)
+                    shutil.rmtree(working_src_dir)
+                    need_to_clone = True
+                working_src_dir.parent.mkdir(parents=True, exist_ok=True)
+            logger.info("%s: moving cloned repo to %s", req.name, working_src_dir)
+            shutil.move(clone_dir, working_src_dir)
+
+    # We must know the version and we must have a source directory.
+    assert version
+    assert working_src_dir
+    assert working_src_dir.exists()
+    return (working_src_dir, version)
+
+
+def _get_version_from_package_metadata(
+    ctx: context.WorkContext,
+    req: Requirement,
+    source_dir: str,
+) -> Version:
+    logger.info(f"{req.name}: generating metadata to get version")
+    pyproject_toml = dependencies.get_pyproject_contents(source_dir)
+    hook_caller = dependencies.get_build_backend_hook_caller(
+        source_dir,
+        pyproject_toml,
+        {},
+        network_isolation=ctx.network_isolation,
+    )
+    metadata_dir_base = hook_caller.prepare_metadata_for_build_wheel(
+        metadata_directory=source_dir.parent,
+        config_settings={},
+    )
+    metadata_filename = source_dir.parent / metadata_dir_base / "METADATA"
+    with open(metadata_filename, "rb") as f:
+        p = BytesParser()
+        metadata = p.parse(f, headersonly=True)
```

Comment:
packaging has code to parse metadata: https://packaging.pypa.io/en/stable/metadata.html

Reply:
Huh, I copied this from some code I found that _you_ wrote. I'll update it.

Reply:
I don't remember why I did not use packaging here...

Reply:
How about if I file a ticket to fix both places and we take this as it is for now?

Reply:
SGTM!

Reply:
#561 tracks this